src/main/java/uk/gov/pay/products/service/ProductApiTokenManager.java

```diff
@@ -0,0 +1,82 @@
+
+public class ProductApiTokenManager {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private Client client;
+    private final String publicAuthUrl;
+    private final String emailAddressForRotatingApiTokens;
+    private final ProductFactory productFactory;
+
+    @Inject
+    public ProductApiTokenManager(Client client, ProductsConfiguration configuration, ProductFactory productFactory) {
+        this.client = client;
+        this.emailAddressForRotatingApiTokens = configuration.getEmailAddressForRotatingApiTokens();
+        this.publicAuthUrl = configuration.getPublicAuthUrl();
+        this.productFactory = productFactory;
+    }
+
+    public String getNewApiTokenFromPublicAuth(Product product) {
+        CreateApiTokenRequest tokenRequest = createApiTokenRequest(product);
+
+        Response response = client.target(publicAuthUrl)
+                .request()
+                .post(Entity.entity(tokenRequest, MediaType.APPLICATION_JSON));
+
+        return Optional.ofNullable(response.readEntity(NewApiTokenFromPublicAuthResponse.class))
+                .map(NewApiTokenFromPublicAuthResponse::getToken)
+                .orElseThrow(() -> new MissingApiTokenException(
+                        String.format("Failed to get a new API token for product %s of type %s", 
+                                product.getExternalId(),
+                                product.getType())));
+    }
+
+    public void rotateApiTokenForAProduct(Product product, String newApiToken) {
+        productFactory.productFinder().updatePayApiTokenByExternalId(product.getExternalId(), newApiToken).ifPresentOrElse(modifiedProduct -> 
+                        logger.info(String.format("Regenerated API token for product %s of type %s", modifiedProduct.getExternalId(), modifiedProduct.getType())),
+                () -> { throw new FailToRotateApiTokenException(
+                        String.format("Failed to rotate API token for product %s of type %s", product.getExternalId(), product.getType())); });
+    }
+
+    private CreateApiTokenRequest createApiTokenRequest(Product product) {
+        return new CreateApiTokenRequest(
+                String.valueOf(product.getGatewayAccountId()),
+                getDescription(product),
+                emailAddressForRotatingApiTokens,
+                CARD,
+                PRODUCTS);
+    }
+
+    private String getDescription(Product product) {
+        switch (product.getType()) {
+            case ADHOC:
+                return String.format("Token for \"%s\" payment link", product.getName());
+            case PROTOTYPE:
+                return String.format("Token for Prototype: %s", product.getName());
+            case DEMO:
+                return "Token for Demo Payment";
+            default:
+                throw new InvalidProductTypeException(String.format("Invalid product type [%s]", product.getType().toString()));
```

Comment:
I think having a specific `InvalidProductTypeException` for something that should never happen is probably overkill. It could just be an `IllegalArgumentException`.